setwd("/Users/LenovoX1/Documents/Timema_facultative_partheno/Analyses/")

######################
### Load libraries ###
######################
library(vcfR)
library(adegenet)
library(car)
library(multcomp)
library(phytools)
library(ridgeline)


########################
### Custom functions ###
########################

'%!in%' <- Negate('%in%')

count.na<-function(x) sum(is.na(x))

# computes the proportion of heterozygous sites out of all genotyped sites
# from a vector of genotypes where genotypes are coded as "0/0", "0/1", or "1/1".
# This format is generated by the function extract.gt(vcf, element = "GT") of the package vcfR.
prop.het<-function(vector) {
  return(sum(vector=="0/1" | vector=="1/0", na.rm=T)/sum(is.na(vector)==F))
}

# This function divides eachvalue in a vector by the average value of that vector.
standardize <- function(vector) return(vector / mean(vector, na.rm=T))

################################
### Read and format the data ###
################################
vcf <- read.vcfR("DP11_miss75_mac3_1snp_all.recode.vcf")

# Extract the genotypes
gt <- extract.gt(vcf, element = "GT")

# Extract depth info for each genotype
dp <- extract.gt(vcf, "DP", as.numeric = T)
# Remove depth info at SNPs where a SNP wasn't called
dp[which(dp < 11)] <- NA
# Standardize depth per individual
rel_dp <- apply(dp, 2, standardize)


X_linked <- read.csv("Tdi_v8_1000_peakadj_2_chr_class.csv", h=T)
autosomal_scaffolds <- X_linked$scaf_name[which(X_linked$chr_type == "A")]
X_scaffolds <- X_linked$scaf_name[which(X_linked$chr_type == "X")]

vcf_X <- vcf[which(vcf@fix[,1] %in% X_scaffolds),]
gt_X <- extract.gt(vcf_X, "GT")
#dp_X <- extract.gt(vcf_X, "DP", as.numeric=T)
gen_X <- vcfR2genlight(vcf_X)
rel_dp_X <- rel_dp[which(vcf@fix[,1] %in% X_scaffolds),]

# Read the metadata, which contains the name of each individual,
# its family, population of origin, and where colour and point shape will be stored
# meta$ID should be identical to dimnames(gt)[[2]]
meta.all <- read.table("meta_all_indiv.csv", sep=";", h=T)

# This extra step ensures that individuals are in the same order 
# in the metadata as in the genotype files.
meta <- meta.all[match(dimnames(gt)[[2]], meta.all$ID),]
meta$ID <- as.character(meta$ID)
meta$subpop <- as.character(meta$subpop)

# Families are named after the mother's name.
meta$family <- NA
meta$family[which(substr(meta$ID, 1, 3) == "17-")] <- substr(meta$ID, 1, 7)

# Computes the percentage of missing data per individual
meta$percent_missing<-apply(gt, 2, count.na)/dim(gt)[[1]] * 100

# Defines different point shapes for different types of individuals
meta$pch=3
meta$pch[which(meta$type=="F")]<-15
meta$pch[which(meta$type=="M")]<-17
meta$pch[which(meta$type=="U")]<-4
meta$pch[which(meta$type=="A")]<-20

meta$col.type <- "black"
meta$col.type[which(meta$type == "U")] <- rgb(0,0.5,1,1)
meta$col.type[which(meta$type == "A")] <- rgb(1,0,1,1)

adults <- read.table("keep_adults.txt")$V1
#Loading the results from faststructure
MO2 <- read.table("structure_DP11_miss75_mac3_1snp_adults_K.2.meanQ")
row.names(MO2) <- adults
MO3 <- read.table("structure_DP11_miss75_mac3_1snp_adults_K.3.meanQ")
row.names(MO3) <- adults
MO4 <- read.table("structure_DP11_miss75_mac3_1snp_adults_K.4.meanQ")
row.names(MO4) <- adults
MO5 <- read.table("structure_DP11_miss75_mac3_1snp_adults_K.5.meanQ")
row.names(MO5) <- adults
MO6 <- read.table("structure_DP11_miss75_mac3_1snp_adults_K.6.meanQ")
row.names(MO6) <- adults

# Loading the sex ratio results
sexratio <- read.table("sex_ratios.txt", h=T)

# Loading the hatching success data for other species
allsp.raw <- read.table("allspp_hatching_unfert.csv", sep=",", h=T)
# Keeping only clutches with at least five eggs
allsp <- allsp.raw[which(allsp.raw$N_eggs >= 5),]

# Loading the metadata for the reference T. douglasi and T. poppensis individuals
treemeta <- read.table("metadata_additional_Tps_Tdi_refs.txt", h=T)

# Loading the nuclear tree (Figure 1E)
tree_nucl <- read.tree("../phylo_mine/subset_mt_DP8_mac3_miss80.recode.min4.phy.treefile.newick")

# Loading the mitochdonrial tree (Figure 1F) and removing the outgroup
tree_mt <- read.tree("../mtDNA_phylo/all_consensus_withoutgroup_aligned.nexus")
tree_mt2 <- drop.tip(tree_mt, tip = "Timema_cristinae", trim.internal = T)

#############################
### Population sex ratios ###
#############################

### Figure 1A & 1B
# We drew the maps and placed the piecharts manually in Illustrator.
for (i in 1:length(sexratio$N)) {
  svg(filename = paste("Figures/piecharts/", sexratio$Pop[i], ".svg", sep=""), width = 4, height = 4)
  pie(c(sexratio$Females[i], sexratio$Males[i]), labels = NA)
  dev.off()
}

# Logistic regressions to test for differences between populations within transect
sex <- integer(0)
pop <- character(0)
for (i in 1:length(sexratio$N)){
  sex <- c(sex, rep(1, sexratio$Females[i]), rep(0, sexratio$Males[i]))
  pop <- c(pop, rep(sexratio$Pop[i], sexratio$N[i]))
}

transect <- substr(pop, 1, 1)
summary(glm(sex[which(transect=="M")]~pop[which(transect=="M")], family="binomial"))
plot(glm(sex[which(transect=="M")]~pop[which(transect=="M")], family="binomial"))
Anova(glm(sex[which(transect=="M")]~pop[which(transect=="M")], family="binomial"))

summary(glm(sex[which(transect=="O")]~pop[which(transect=="O")], family="binomial"))
plot(glm(sex[which(transect=="O")]~pop[which(transect=="O")], family="binomial"))
Anova(glm(sex[which(transect=="O")]~pop[which(transect=="O")], family="binomial"))


######################
### Heterozygosity ###
######################

meta$het <- apply(gt, 2, prop.het)

### Figure S1
par(mfrow=c(1,1), mai=c(0.8, 0.8, 0.1, 0.1))
with(meta[which(meta$type=="U"),], plot(het ~ percent_missing, las=1, xlab="Percentage of missing data", ylim=c(0, 0.1), ylab="Relative heterozygosity"))
lm_het <- with(meta[which(meta$type=="U"),], lm_het <- lm(het~percent_missing))
summary(lm_het)
abline(lm_het, lty=2)
abline(h=lm_het$coefficients[1], lty=3)

# Defining the corrected value of heterozygosity as 
# the observed value minus the difference between 
# the predicted value and the intercept.
meta$het.corrected <- meta$het - predict.lm(object = lm_het, newdata = meta) + lm_het$coefficients[1]

# Adding the mother's heterozygosity value to each offspring
# (will be used for ordering in Figure 3)
meta$mom.het <- NA
for (mom in meta$ID[which(meta$type=="F")]) {
  meta$mom.het[which(substr(meta$ID, 1, 7)==mom)]<-meta$het.corrected[which(meta$ID==mom)]
  meta$mom.het[which(meta$type=="F")]<-NA
}

############################
### Population structure ###
############################

# Sorting individuals per population
orderMO <- c(
  meta$ID[which(meta$subpop == "MANCH_1" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_2" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_3" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_4" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_5" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_6" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_7" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_10" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "MANCH_12" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_1" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_3" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_4" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_5" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_8" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_11" & meta$ID %in% adults)],
  meta$ID[which(meta$subpop == "ORR_15" & meta$ID %in% adults)]
)


# Figure S1
par(mfrow=c(6,1), mai=c(0.4, 0.4, 0.1, 0.1))
barplot(meta$het.corrected[match(orderMO, meta$ID)], border=NULL, las=1, ylim=c(0,0.3))
barplot(t(as.matrix(MO2))[,match(orderMO, dimnames(t(as.matrix(MO2)))[[2]])], las=2, col=funky(2), border=NA, xaxt="n", yaxp=c(0,1,2))
barplot(t(as.matrix(MO3))[,match(orderMO, dimnames(t(as.matrix(MO3)))[[2]])], las=2, col=c("#841433", "#98D9EB", "#FEDA00"), border=NA, xaxt="n", yaxp=c(0,1,2))
barplot(t(as.matrix(MO4))[,match(orderMO, dimnames(t(as.matrix(MO4)))[[2]])], las=2, col=funky(4), border=NA, xaxt="n", yaxp=c(0,1,2))
barplot(t(as.matrix(MO5))[,match(orderMO, dimnames(t(as.matrix(MO5)))[[2]])], las=2, col=funky(5), border=NA, xaxt="n", yaxp=c(0,1,2))
barplot(t(as.matrix(MO6))[,match(orderMO, dimnames(t(as.matrix(MO6)))[[2]])], las=2, col=funky(6), border=NA, cex.names=0.8, yaxp=c(0,1,2))

### Figure 1C
par(mfrow=c(1,1))
barplot(t(as.matrix(MO3))[,match(orderMO, dimnames(t(as.matrix(MO3)))[[2]])], las=2, col=c("#841433", "#98D9EB", "#FEDA00"), border=NA, xaxt="n", yaxp=c(0,1,2))

# Define colours for the other figures
meta$colK3 <- "grey"
meta$lineage <- NA
meta$colK3[match(row.names(MO3)[which(MO3$V3 > 0.8)], meta$ID)] <- "#98D9EB"
meta$lineage[match(row.names(MO3)[which(MO3$V3 > 0.8)], meta$ID)] <- "blue"
meta$colK3[match(row.names(MO3)[which(MO3$V1 > 0.8)], meta$ID)] <- "#FEDA00"
meta$lineage[match(row.names(MO3)[which(MO3$V1 > 0.8)], meta$ID)] <- "yellow"
meta$colK3[match(row.names(MO3)[which(MO3$V2 > 0.8)], meta$ID)] <- "#841433"
meta$lineage[match(row.names(MO3)[which(MO3$V2 > 0.8)], meta$ID)] <- "red"
meta$lineage <- as.factor(meta$lineage)

#################
### Phylogeny ###
#################
cophylo <- cophylo(tree_nucl, tree_mt2)
cophylo$trees[[1]]$col <- meta$colK3[match(cophylo$trees[[1]]$tip.label, meta$ID)]
cophylo$trees[[1]]$col[which(cophylo$trees[[1]]$tip.label %in% treemeta$ID[which(treemeta$sp == "poppensis")])] <- "#00A79D"
cophylo$trees[[1]]$col[which(cophylo$trees[[1]]$tip.label %in% treemeta$ID[which(treemeta$sp == "douglasi")])] <- "#662D91"
cophylo$trees[[2]]$col <- meta$colK3[match(cophylo$trees[[2]]$tip.label, meta$ID)]
cophylo$trees[[2]]$col[which(cophylo$trees[[2]]$tip.label %in% treemeta$ID[which(treemeta$sp == "poppensis")])] <- "#00A79D"
cophylo$trees[[2]]$col[which(cophylo$trees[[2]]$tip.label %in% treemeta$ID[which(treemeta$sp == "douglasi")])] <- "#662D91"

plot(cophylo, fsize=0.45, scale.bar=c(0.1, 0.05))
tiplabels.cophylo(pch=19, col=cophylo$trees[[1]]$col, which="left")
tiplabels.cophylo(pch=19, col=cophylo$trees[[2]]$col, which="right")

#########################
### Reproductive mode ###
#########################

### Reproductive mode differences between the lineages

meta$hatch_ratio_unf <- meta$N_hatching_unf  / meta$N_unfertilized_eggs
meta$hatch_ratio_unf[which(is.nan(meta$hatch_ratio_unf))] <- NA
meta$hatch_ratio_after <- meta$N_hatching_aftermating / meta$N_eggs_aftermating
meta$total_eggs <- meta$N_unfertilized_eggs + meta$N_eggs_aftermating

# Removing 3 females for which no offspring could be obtained
moms <- meta$ID[which(meta$type=="F")]
moms <- moms[-which(moms %in% c("17-2161", "17-2204", "17-2009"))]

### Figure 2A
meta$order <- NA
meta$order[which(meta$subpop == "MANCH_1")] <- 1
meta$order[which(meta$subpop == "MANCH_2")] <- 2
meta$order[which(meta$subpop == "MANCH_3")] <- 3
meta$order[which(meta$subpop == "MANCH_4")] <- 4
meta$order[which(meta$subpop == "MANCH_5")] <- 5
meta$order[which(meta$subpop == "MANCH_6")] <- 6
meta$order[which(meta$subpop == "MANCH_7")] <- 7
meta$order[which(meta$subpop == "MANCH_8")] <- 8
meta$order[which(meta$subpop == "MANCH_9")] <- 9
meta$order[which(meta$subpop == "MANCH_10")] <- 10
meta$order[which(meta$subpop == "MANCH_11")] <- 11
meta$order[which(meta$subpop == "MANCH_12")] <- 12
meta$order[which(meta$subpop == "ORR_1")] <- 13
meta$order[which(meta$subpop == "ORR_2")] <- 14
meta$order[which(meta$subpop == "ORR_3")] <- 15
meta$order[which(meta$subpop == "ORR_4")] <- 16
meta$order[which(meta$subpop == "ORR_7")] <- 17
meta$order[which(meta$subpop == "ORR_8")] <- 18
meta$order[which(meta$subpop == "ORR_10")] <- 19
meta$order[which(meta$subpop == "ORR_11")] <- 20
meta$order[which(meta$subpop == "ORR_17")] <- 21
ordered_pops <- c("MANCH_1", "MANCH_2", "MANCH_3", "MANCH_4", "MANCH_5", "MANCH_6", "MANCH_7", "MANCH_8", "MANCH_9", "MANCH_10", "MANCH_11", "MANCH_12", "ORR_1", "ORR_2", "ORR_3", "ORR_4", "ORR_7", "ORR_8", "ORR_10", "ORR_11")

boxplot(meta$hatch_ratio_unf ~ meta$order, col=NA, las=2, outline=F, xlab="Population", ylab="Proportion of unfertilized eggs hatching", names=ordered_pops, ylim=c(0,1))
points(meta$hatch_ratio_unf ~ jitter(eggs$order), col=eggs$col, pch=19, cex=sqrt(log(eggs$N_unf2/3)))
legend("topleft", legend=c("5", "15", "60"), pch=19, cex=sqrt(log(c(5,15,60)/3)), bty="n")


# Figure 2B
ourdata <- data.frame(ID = eggs$ID, sp = eggs$lineage, year = "2017", location=eggs$subpop, N_eggs=eggs$N_unf, N_hatched=eggs$N_hatch_unf, hatch_ratio=eggs$hatch_ratio_unf)
ourdata$sp[which(ourdata$sp == 1)] <- "Western Manchester"
ourdata$sp[which(ourdata$sp == 2)] <- "Eastern Manchester"
ourdata$sp[which(ourdata$sp == 3)] <- "Orr"

allsp$hatch_ratio <- allsp$N_hatched / allsp$N_eggs
ridgeline(allsp$hatch_ratio, allsp$sp, bw=0.01, border =  NA)

all_unf.raw <- rbind(allsp, ourdata)
all_unf <- all_unf.raw[which(all_unf.raw$N_eggs >= 5),]

all_unf$sp[which(all_unf$sp == "petita")] <- "01_T. petita"
all_unf$sp[which(all_unf$sp == "podura")] <- "02_T. podura"
all_unf$sp[which(all_unf$sp == "cristinae")] <- "03_T. cristinae"
all_unf$sp[which(all_unf$sp == "chumash")] <- "04_T. chumash"
all_unf$sp[which(all_unf$sp == "knulli")] <- "05_T. knulli"
all_unf$sp[which(all_unf$sp == "californicum")] <- "06_T. californicum"
all_unf$sp[which(all_unf$sp == "poppensis")] <- "07_T. poppensis"
all_unf$sp[which(all_unf$sp == "Western Manchester")] <- "08_Western_Manchester"
all_unf$sp[which(all_unf$sp == "Eastern Manchester")] <- "09_Eastern_Manchester"
all_unf$sp[which(all_unf$sp == "Orr")] <- "10_Orr"
all_unf$sp[which(all_unf$sp == "douglasi")] <- "11_douglasi"
all_unf$sp[which(all_unf$sp == "shepardi")] <- "12_shepardi"
all_unf$sp[which(all_unf$sp == "monikensis")] <- "13_monikensis"
all_unf$sp[which(all_unf$sp == "genevievae")] <- "14_genevievae"
all_unf$sp[which(all_unf$sp == "tahoe")] <- "15_tahoe"

hist(all_unf$hatch_ratio[which(all_unf$sp %!in% c("08_Western_Manchester", "09_Eastern_Manchester", "10_Orr"))], breaks=100, main="", las=1)
hist(all_unf$hatch_ratio, breaks=100, main="", las=1)


par(mai=c(0.5, 1.8, 0.1, 0.1))
ridgeline(all_unf$hatch_ratio, all_unf$sp, bw=0.03, palette = c(hcl.colors(7, palette = "viridis", alpha = 1), "#841433", "#FEDA00", "#98D9EB"))


### Figure 3

# ordering mothers by genetic lineage, then by increasing heterozygosity
order <- c(meta$ID[which(meta$type=="F" & meta$colK3== "#98D9EB")][order(meta$het.corrected[which(meta$type=="F" & meta$colK3== "#98D9EB")])], meta$ID[which(meta$type=="F" & meta$colK3== "#FEDA00")][order(meta$het.corrected[which(meta$type=="F" & meta$colK3== "#FEDA00")])], meta$ID[which(meta$type=="F" & meta$colK3== "#841433")][order(meta$het.corrected[which(meta$type=="F" & meta$colK3== "#841433")])], "17-2162")
# manually removing all mothers without unfertilized offspring
order <- order[-which(order %in% c("17-2161", "17-1913", "17-1937", "17-1800", "17-2204", "17-1796", "17-1770", "17-2009", "17-1797"))] 

par(mfrow=c(1,1), mai=c(0.8,0.8, 0.1, 0.1))
plot(meta$het.corrected[match(order, meta$ID)], pch=19, cex=1.25, col=meta$colK3[match(order, meta$ID)], xaxt="n", xlab="", ylab="Relative_heterozygosity", las=1, ylim=c(0, max(meta$het[which(meta$ID %in% order)])))
axis(side=1, at=1:27, labels = order, las=2, cex.axis=0.8)
points(meta$het.corrected[which(meta$type=="U")] ~ jitter(match(meta$family[which(meta$type=="U")], order)), pch=4, cex=0.75)


### Calculating the proporiton of transitions 
### from homozygous in the mother to heterozygous in the offspring
### (called mismatches_mith_mom in this script)
meta$homo_in_mom <- NA
meta$mismatches_with_mom <- NA
meta$prop_mismatches_with_mom <- NA
for (baby in meta$ID[which(meta$type %in% c("A", "U"))]) {
  mom <- meta$family[which(meta$ID == baby)]
  baby.gt <- gt[,which(dimnames(gt)[[2]]==baby)]
  mom.gt <- gt[,which(dimnames(gt)[[2]]==mom)]
  genotyped.in.both <- which(is.na(baby.gt)==F & is.na(mom.gt)==F)
  homo_in_mom <- which(mom.gt[genotyped.in.both] %in% c("0/0", "1/1"))
  meta$homo_in_mom[which(meta$ID == baby)] <- length(homo_in_mom)
  mismatches_with_mom <- which(baby.gt[homo_in_mom] == "0/1" )
  meta$mismatches_with_mom[which(meta$ID == baby)] <- length(mismatches_with_mom)
  meta$prop_mismatches_with_mom[which(meta$ID == baby)] <- meta$mismatches_with_mom[which(meta$ID == baby)] / meta$homo_in_mom[which(meta$ID == baby)]
}



### Figure 4
plot(meta$prop_mismatches_with_mom[which(meta$type %in% c("A", "U"))] ~ meta$het.corrected[which(meta$type %in% c("A", "U"))], col=meta$col.type[which(meta$type %in% c("A", "U"))], las=1, ylab = "Proportion of heterozygous transitions", xlab="Corrected heterozygosity", pch=meta$pch[which(meta$type %in% c("A", "U"))], xlim=c(0,0.3), ylim=c(0,0.3))

# We define individuals with corrected heterozygosity < 0.1 
# & prop_mismatches_with_mom < 0.08 as produced via parthenogenesis
# others are assumed to be produced by sex.
meta$produced_by <- NA
meta$produced_by[which(meta$prop_mismatches_with_mom > 0.1 & meta$het > 0.08)] <- "sex"
meta$produced_by[which(meta$prop_mismatches_with_mom < 0.1 & meta$het < 0.08)] <- "asex"


# Computing the proportion of babies from eggs laid after mating that were produced via sex (i.e. fertilized).
meta$prop_after_fert <- NA
meta$N_after_fert <- NA
meta$N_after_unfert <- NA
for (i in moms) {
  if (meta$hatch_ratio_after[which(meta$ID == i)] > 0) {
    fert <- sum(meta$family == i & meta$type=="A" & meta$produced_by =="sex", na.rm=T)
    unfert <- sum(meta$family == i & meta$type=="A" & meta$produced_by =="asex", na.rm=T)
    meta$N_after_fert[which(meta$ID==i)] <- fert
    meta$N_after_unfert[which(meta$ID==i)] <- unfert
    meta$prop_after_fert[which(meta$ID==i)] <- fert/(fert+unfert)
  }
}
meta$prop_after_fert[which(is.nan(meta$prop_after_fert))] <- NA

meta$N_genotyped_after <- 0
meta$N_genotyped_after[match(names(table(meta$family[which(meta$type=="A")])), meta$ID)] <- as.numeric(table(meta$family[which(meta$type=="A")]))

# Difference in hatching success of unfertilized eggs between lineages
# (Results presented in Figure 5A)
glm1 <- glm(cbind(meta$N_hatching_unf, (meta$N_unfertilized_eggs - meta$N_hatching_unf)) ~ as.factor(meta$lineage), family=quasibinomial)
summary(glm1)
anova(glm1)
plot(glm1)
summary(glht(glm1, mcp("as.factor(meta$lineage)"="Tukey")))
predicted.glm1 <- predict.glm(glm1, se.fit = T, type="response")


# Difference in hatching success of eggs laid after mating between lineages
# (Results presented in Figure 5B)
glm2 <- glm(cbind(meta$N_hatching_aftermating, (meta$N_eggs_aftermating-meta$N_hatching_aftermating)) ~ as.factor(meta$lineage), family=quasibinomial)
summary(glm2)
anova(glm2)
plot(glm2)
summary(glht(glm2, mcp("as.factor(meta$lineage)"="Tukey")))
(predicted.glm2 <- predict.glm(glm2, se.fit = T, type="response"))



# Difference in proportion of eggs laid after mating that are fertilized between lineages
# (Results presented in Figure 5C)
glm3 <- glm(cbind(meta$N_after_fert, meta$N_after_unfert) ~ as.factor(meta$lineage), family=quasibinomial)
summary(glm3)
anova(glm3)
plot(glm3)
summary(glht(glm3, mcp("as.factor(meta$lineage)"="Tukey")))
(predicted.glm3 <- predict.glm(glm3, se.fit = T, type="response"))

### Figure 5
# to order individuals by lineage
meta$lineageN <- NA
meta$lineageN[which(meta$lineage=="red")] <- 1
meta$lineageN[which(meta$lineage=="yellow")] <- 2
meta$lineageN[which(meta$lineage=="blue")] <- 3


par(mfrow=c(1,3))

# Values below were copied by hand because I'm lazy but were taken
#Figure 5A
plot(meta$hatch_ratio_unf ~ jitter(meta$lineageN, amount = 0.1), cex=sqrt(meta$N_unfertilized_eggs+1)/4,  pch=19, col=meta$colK3, las=1, ylab="Hatching success of virgin eggs", xlab="Genetic lineage")
points(x=c(1, 2, 3), y=c(0.4213483, 0.3287037, 0.6682692), pch=15, cex=1.5)
segments(x0=1, x1=1, y0=0.4213483 - 0.07185943, y1=0.4213483 + 0.07185943, lwd=2)
segments(x0=2, x1=2, y0=0.3287037 - 0.06205789, y1=0.3287037 + 0.06205789, lwd=2)
segments(x0=3, x1=3, y0=0.6682692 - 0.06338718, y1=0.6682692 + 0.06338718, lwd=2)

# Figure 5B
plot(meta$hatch_ratio_after ~ jitter(meta$lineageN, amount = 0.2), col=meta$colK3, cex=sqrt(meta$N_eggs_aftermating)/4, pch=19, las=1, ylab="Hatching success of eggs laid after mating", xlab="Genetic lineage")
points(x=c(1, 2, 3), y=c(0.5526316, 0.5495495 , 0.5580524), pch=15, cex=1.5)
segments(x0=1, x1=1, y0=0.5526316 - 0.08809832, y1=0.5526316 + 0.08809832, lwd=2)
segments(x0=2, x1=2, y0=0.5495495 - 0.06317133, y1=0.5495495 + 0.06317133, lwd=2)
segments(x0=3, x1=3, y0=0.5580524 - 0.08131147, y1=0.5580524 + 0.08131147, lwd=2)

# Figure 5C
plot(meta$prop_after_fert ~ jitter(meta$lineageN, amount = 0.2), col=meta$colK3, cex=sqrt(meta$N_unfertilized_eggs)/4, pch=19, las=1, ylab="Proportion of fertilized eggs after mating", xlab="Genetic lineage")
points(x=c(1, 2, 3), y=c(0.9466667, 0.7898089 , 0.1034483), pch=15, cex=1.5)
segments(x0=1, x1=1, y0=0.9466667 - 0.06252173, y1=0.9466667 + 0.06252173, lwd=2)
segments(x0=2, x1=2, y0=0.7898089 - 0.07835778, y1=0.7898089 + 0.07835778, lwd=2)
segments(x0=3, x1=3, y0=0.1034483 - 0.06813701, y1=0.1034483 + 0.06813701, lwd=2)

###############################################################
### Leftover heterozygosity in asexually-produced offspring ###
###############################################################

meta$het_X <- apply(gt_X, 2, prop.het)
male_X_het <- apply(gt_X[,which(dimnames(gt)[[2]] %in% meta$ID[which(meta$type=="M")])], 2, prop.het)
unf_X_het <- apply(gt_X[,which(dimnames(gt)[[2]] %in% meta$ID[which(meta$produced_by=="asex")])], 2, prop.het)
unf_het <- apply(gt[,which(dimnames(gt)[[2]] %in% meta$ID[which(meta$produced_by=="asex")])], 2, prop.het)

### using the same correction as in the main text
lm_het_male_X <- with(meta[which(meta$type=="M"),], lm_het_male_X <- lm(het_X~percent_missing))
meta$het_male_X.corrected <- NA
meta$het_male_X.corrected[which(meta$type=="M")] <- male_X_het - predict.lm(object = lm_het_male_X, newdata = meta[which(meta$type=="M"),]) + lm_het_male_X$coefficients[1]

lm_het_unf_X <- with(meta[which(meta$produced_by=="asex"),], lm_het_unf_X <- lm(het_X~percent_missing))
meta$het_unf_X.corrected <- NA
meta$het_unf_X.corrected[which(meta$produced_by=="asex")] <- unf_X_het - predict.lm(object = lm_het_unf_X, newdata = meta[which(meta$produced_by=="asex"),]) + lm_het_unf_X$coefficients[1]

### Figure S4A
boxplot(meta$het_male_X.corrected, meta$het_unf_X.corrected, meta$het.corrected[which(meta$produced_by=="asex")], names = c("Male X", "X of asex babies", "asex babies genomewide"), las=1, ylab="Relative heterozygosity", col=NULL, outline = F, ylim=c(0, max(meta$het_male_X.corrected, na.rm = T)))
points(meta$het_male_X.corrected ~ jitter(rep(1, dim(gt)[2]), factor = 12), pch=19)
points(meta$het_unf_X.corrected ~ jitter(rep(2, dim(gt)[2]), factor = 8), pch=19)
points(meta$het.corrected[which(meta$produced_by=="asex")] ~ jitter(rep(3, sum(meta$produced_by=="asex", na.rm = T)), factor = 5), pch=19)

t.test(meta$het_unf_X.corrected, meta$het_male_X.corrected)

### Depth comparison at homo- vs heterozygous loci

avg_rel_dp_het_male_X <- apply(rel_dp_X[which(X_het_prop_in_males == 1), which(dimnames(gt)[[2]] %in% meta$ID[which(meta$type=="M")])], 2, mean, na.rm=T)
avg_rel_dp_homo_male_X <- apply(rel_dp_X[which(X_het_prop_in_males == 0), which(dimnames(gt)[[2]] %in% meta$ID[which(meta$type=="M")])], 2, mean, na.rm=T)

### Figure S4B
par(mfrow=c(1,1))
boxplot(avg_rel_dp_homo_male_X, avg_rel_dp_het_male_X, names=c("Homozygous", "Heterozygous"), las=1, ylab="Average standardized depth", col=NULL, outline=F, ylim=c(0,max(avg_rel_dp_het_male_X)))
x1 <- jitter(rep(1, length(avg_rel_dp_het_male_X)), 10)
x2 <- x1 + 1
points(avg_rel_dp_homo_male_X ~ x1, pch=19)
points(avg_rel_dp_het_male_X ~ x2, pch=19)
segments(x1, avg_rel_dp_homo_male_X, x2, avg_rel_dp_het_male_X)

t.test(avg_rel_dp_homo_male_X, avg_rel_dp_het_male_X, paired = T)

mean(avg_rel_dp_het_male_X) / mean(avg_rel_dp_homo_male_X)
